coderd/database/queries/prebuilds.sql

```diff
@@ -0,0 +1,141 @@
+-- We also return the number of failed workspace builds that occurred during the lookback period.
+--
+-- NOTE:
+-- - To **decide whether to back off**, we look at up to the N most recent builds (within the defined lookback period).
+-- - To **calculate the number of failed builds**, we consider all builds within the defined lookback period.
+--
+-- The number of failed builds is used downstream to determine the backoff duration.
+-- name: GetPresetsBackoff :many
+WITH filtered_builds AS (
+	-- Only select builds which are for prebuild creations
+	SELECT wlb.template_version_id, wlb.created_at, tvp.id AS preset_id, wlb.job_status, tvp.desired_instances
+	FROM template_version_presets tvp
+			 INNER JOIN workspace_latest_builds wlb ON wlb.template_version_preset_id = tvp.id
+             INNER JOIN template_versions tv ON wlb.template_version_id = tv.id
+             INNER JOIN templates t ON tv.template_id = t.id AND t.active_version_id = tv.id
+	WHERE tvp.desired_instances IS NOT NULL -- Consider only presets that have a prebuild configuration.
+      AND wlb.transition = 'start'::workspace_transition
+),
+time_sorted_builds AS (
+    -- Group builds by template version, then sort each group by created_at.
+	SELECT fb.template_version_id, fb.created_at, fb.preset_id, fb.job_status, fb.desired_instances,
+	    ROW_NUMBER() OVER (PARTITION BY fb.preset_id ORDER BY fb.created_at DESC) as rn
+	FROM filtered_builds fb
+),
+failed_count AS (
+    -- Count failed builds per template version/preset in the given period
+	SELECT preset_id, COUNT(*) AS num_failed
+	FROM filtered_builds
+	WHERE job_status = 'failed'::provisioner_job_status
+		AND created_at >= @lookback::timestamptz
+	GROUP BY preset_id
+)
+SELECT tsb.template_version_id,
+	   tsb.preset_id,
+	   COALESCE(fc.num_failed, 0)::int  AS num_failed,
+	   MAX(tsb.created_at)::timestamptz AS last_build_at
+FROM time_sorted_builds tsb
+		 LEFT JOIN failed_count fc ON fc.preset_id = tsb.preset_id
+WHERE tsb.rn <= tsb.desired_instances -- Fetch the last N builds, where N is the number of desired instances; if any fail, we backoff
+  AND tsb.job_status = 'failed'::provisioner_job_status
+  AND created_at >= @lookback::timestamptz
+GROUP BY tsb.template_version_id, tsb.preset_id, fc.num_failed;
+
+-- name: ClaimPrebuild :one
+UPDATE workspaces w
+SET owner_id   = @new_user_id::uuid,
+	name       = @new_name::text,
+	updated_at = NOW()
+WHERE w.id IN (
+	SELECT p.id
+	FROM workspace_prebuilds p
+		INNER JOIN workspace_latest_builds b ON b.workspace_id = p.id
+		INNER JOIN templates t ON p.template_id = t.id
+	WHERE (b.transition = 'start'::workspace_transition
+		AND b.job_status IN ('succeeded'::provisioner_job_status))
+	-- The prebuilds system should never try to claim a prebuild for an inactive template version.
+	-- Nevertheless, this filter is here as a defensive measure:
+	AND b.template_version_id = t.active_version_id
+	AND b.template_version_preset_id = @preset_id::uuid
```

Comment:
```suggestion
	AND cp.current_preset_id = @preset_id::uuid
```

this would allow you to claim a prebuilt workspace that had been rebuilt.

Reply:
You mean `p.current_preset_id`?
I think it make sense - done.